File: front-end/src/app/core/services/search.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface SearchResultDto {
  terme: string;
  totalResultats: number;
  processus: SearchItemDto[];
  procedures: SearchItemDto[];
  documents: SearchItemDto[];
  nonConformites: SearchItemDto[];
  indicateurs: SearchItemDto[];
}

export interface SearchItemDto {
  id: string;
  type: string;
  code: string;
  titre: string;
  description?: string;
  statut?: string;
  dateCreation: string;
}

@Injectable({ providedIn: 'root' })
export class SearchService {

  private readonly API = `${environment.apiUrl}/Search`;

  constructor(private http: HttpClient) {}

  search(terme: string, organisationId: string, options?: {
    includeProcessus?: boolean;
    includeProcedures?: boolean;
    includeDocuments?: boolean;
    includeNonConformites?: boolean;
    includeIndicateurs?: boolean;
  }): Observable<SearchResultDto> {
    let params = new HttpParams()
      .set('terme', terme)
      .set('organisationId', organisationId);

    if (options?.includeProcessus      !== undefined) params = params.set('includeProcessus',      options.includeProcessus.toString());
    if (options?.includeProcedures     !== undefined) params = params.set('includeProcedures',     options.includeProcedures.toString());
    if (options?.includeDocuments      !== undefined) params = params.set('includeDocuments',      options.includeDocuments.toString());
    if (options?.includeNonConformites !== undefined) params = params.set('includeNonConformites', options.includeNonConformites.toString());
    if (options?.includeIndicateurs    !== undefined) params = params.set('includeIndicateurs',    options.includeIndicateurs.toString());

    return this.http.get<SearchResultDto>(this.API, { params });
  }
}

File: back-end/Services/SearchService.cs
using Dapper;
using DocApi.DTOs;
using DocApi.Infrastructure;
using DocApi.Services.Interfaces;

namespace DocApi.Services
{
    public class SearchService : ISearchService
    {
        private readonly IDbConnectionFactory _factory;

        public SearchService(IDbConnectionFactory factory)
        {
            _factory = factory;
        }

        public async Task<SearchResultDto> SearchAsync(SearchRequestDto request)
        {
            using var conn = _factory.CreateConnection();
            var t      = $"%{request.Terme.Trim()}%";
            var orgId  = request.OrganisationId;
            var result = new SearchResultDto { Terme = request.Terme };

            var tasks = new List<Task>();

            if (request.IncludeProcessus)
                tasks.Add(Task.Run(async () =>
                {
                    var rows = await conn.QueryAsync<SearchItemDto>(@"
                        SELECT Id, Code, Nom AS Titre, Description, Statut, DateCreation
                        FROM Processus
                        WHERE OrganisationId = @OrgId
                          AND (Nom LIKE @T OR Code LIKE @T OR Description LIKE @T)
                        LIMIT 10", new { OrgId = orgId, T = t });
                    result.Processus = rows.Select(r => { r.Type = "Processus"; return r; }).ToList();
                }));

            if (request.IncludeProcedures)
                tasks.Add(Task.Run(async () =>
                {
                    var rows = await conn.QueryAsync<SearchItemDto>(@"
                        SELECT Id, Code, Titre, Description, Statut, DateCreation
                        FROM Procedures
                        WHERE OrganisationId = @OrgId
                          AND (Titre LIKE @T OR Code LIKE @T OR Description LIKE @T)
                        LIMIT 10", new { OrgId = orgId, T = t });
                    result.Procedures = rows.Select(r => { r.Type = "Procedure"; return r; }).ToList();
                }));

            if (request.IncludeDocuments)
                tasks.Add(Task.Run(async () =>
                {
                    var rows = await conn.QueryAsync<SearchItemDto>(@"
                        SELECT Id, Code, Titre,
                               Description,
                               CASE WHEN Actif = 1 THEN 'ACTIF' ELSE 'INACTIF' END AS Statut,
                               DateCreation
                        FROM Documents
                        WHERE OrganisationId = @OrgId AND Actif = TRUE
                          AND (Titre LIKE @T OR Code LIKE @T OR Description LIKE @T)
                        LIMIT 10", new { OrgId = orgId, T = t });
                    result.Documents = rows.Select(r => { r.Type = "Document"; return r; }).ToList();
                }));

            if (request.IncludeNonConformites)
                tasks.Add(Task.Run(async () =>
                {
                    var rows = await conn.QueryAsync<SearchItemDto>(@"
                        SELECT Id, Reference AS Code, Description AS Titre,
                               Description, Statut, DateCreation
                        FROM NonConformites
                        WHERE OrganisationId = @OrgId
                          AND (Reference LIKE @T OR Description LIKE @T)
                        LIMIT 10", new { OrgId = orgId, T = t });
                    result.NonConformites = rows.Select(r => { r.Type = "NonConformite"; return r; }).ToList();
                }));

            if (request.IncludeIndicateurs)
                tasks.Add(Task.Run(async () =>
                {
                    var rows = await conn.QueryAsync<SearchItemDto>(@"
                        SELECT Id, Code, Nom AS Titre, Description,
                               CASE WHEN Actif = 1 THEN 'ACTIF' ELSE 'INACTIF' END AS Statut,
                               DateCreation
                        FROM Indicateurs
                        WHERE OrganisationId = @OrgId
                          AND (Nom LIKE @T OR Code LIKE @T OR Description LIKE @T)
                        LIMIT 10", new { OrgId = orgId, T = t });
                    result.Indicateurs = rows.Select(r => { r.Type = "Indicateur"; return r; }).ToList();
                }));

            await Task.WhenAll(tasks);

            result.TotalResultats =
                result.Processus.Count      +
                result.Procedures.Count     +
                result.Documents.Count      +
                result.NonConformites.Count +
                result.Indicateurs.Count;

            return result;
        }
    }
}
